K8SPXC-172 document full data at rest encryption with vault (#362)
percona/percona-xtradb-cluster-operator: source/encryption.rst

Data at rest encryption
************************

`Ful data at rest encryption in Percona XtraDB Cluster <https://www.percona.com/doc/percona-xtradb-cluster/LATEST/management/data_at_rest_encryption.html>`_ is supported by the Operator since version 1.4.0.

..note:: "Data at rest" means inactive data stored as files, database records, etc.

To implement these features, the Operator uses ``keyring_vault`` plugin,
which ships with Percona XtraDB Cluster, and utilizes `HashiCorp Vault <https://www.vaultproject.io/>`_ authentication server. 

.. contents:: :local:

.. _install-vault:

Installing and configuring Vault
-------------------------------

The following steps will deploy Vault on Kubernetes with the `Helm 3 package manager <https://helm.sh/>`_:

1. Clone the official HashiCorp Vault Helm chart from GitHub:

   .. code:: bash

      $ git clone -b v0.4.0 https://github.com/hashicorp/vault-helm.git
      $ cd vault-helm

2. Checkout to a tagged Vault release version to install it with Helm:

   .. code:: bash

>       $ git checkout v0.4.0

Reviewer: not needed, covered point 1.

Data at rest encryption
************************

`Ful data at rest encryption in Percona XtraDB Cluster <https://www.percona.com/doc/percona-xtradb-cluster/LATEST/management/data_at_rest_encryption.html>`_ is supported by the Operator since version 1.4.0.

..note:: "Data at rest" means inactive data stored as files, database records, etc.

To implement these features, the Operator uses ``keyring_vault`` plugin,

> which ships with Percona XtraDB Cluster 8.0, and utilizes `HashiCorp Vault <https://www.vaultproject.io/>`_ authentication server. 

Reviewer: ```suggestion
which ships with Percona XtraDB Cluster, and utilizes `HashiCorp Vault <https://www.vaultproject.io/>`_ authentication server. 
```

.. contents:: :local:

.. _install-vault:

Installing and configuring Vault
-------------------------------

> The following steps will deploy Vault on Kubernetes with the `Helm 3 package manager <https://helm.sh/>`_:

Reviewer: Maybe it is needed to mention that user feel free to setup Vault as he wants, steps 1-4 are only illustrative

1. Clone the official HashiCorp Vault Helm chart from GitHub:

   .. code:: bash

>       $ git clone https://github.com/hashicorp/vault-helm.git

Reviewer: ```suggestion
      $ git clone -b v0.4.0 https://github.com/hashicorp/vault-helm.git
```
please remove step 2.

      $ cd vault-helm

2. Checkout to a tagged Vault release version to install it with Helm:

   .. code:: bash

      $ git checkout v0.4.0

3. Now use Helm to do the installation:

   .. code:: bash

>       $ helm install --name vault-service --namespace vault-namespace ./

Reviewer: ```suggestion
      $ helm install --name vault-service ./
```
let's avoid namespace because next commands expect that it is current namespace

4. After the installation, Vauld should be first initialized and then unsealed.
   Initializing Vault is done with the following commands:

   .. code:: bash

      $ kubectl exec -it pod/vault-service-0 -- vault operator init -key-shares=1 -key-threshold=1 -format=json > /tmp/vault-init
      $ unsealKey=$(jq -r ".unseal_keys_b64[]" < /tmp/vault-init)

>       $ token=$(jq -r ".root_token" < /tmp/vault-init)

Reviewer: we parse token in step 5.
```suggestion
```

   To unseal Vault, execute the following command **for each Pod** of Vault
   running: 

   .. code:: bash

      $ kubectl exec -it pod/vault-service-0 -- vault operator unseal "$unsealKey"

5. Now it is time to enable secrets within Vault. First, get the Vault root
   token:

   .. code:: bash

      $ cat /tmp/vault-init | jq -r ".root_token"

   The output will be like follows:

   .. code:: text

      s.VgQvaXl8xGFO1RUxAPbPbsfN

>    Now login to Vault with this token and enable secrets:
> 
>       $ kubectl exec -it vault-0 -- /bin/sh
>       $ vault login s.VgQvaXl8xGFO1RUxAPbPbsfN
>       $ vault secrets enable --version=1 -path=secret kv

Reviewer: let's use `pxc-secret` path in all examples
```suggestion
   Now login to Vault with this token and enable "pxc-secret" secrets path:

      $ kubectl exec -it vault-0 -- /bin/sh
      $ vault login s.VgQvaXl8xGFO1RUxAPbPbsfN
      $ vault secrets enable --version=1 -path=pxc-secret kv
```

   .. note:: You can also enable audit, which is not mandatory, but useful:

      .. code:: bash

         $ vault audit enable file file_path=/vault/vault-audit.log

> 6. To enable Vault secret within Kubernetes, create and apply the YAML file as

Reviewer: @blez please work with @fiowro to add TLS cert in example

   follows::

      apiVersion: v1
      kind: Secret
      metadata:

>         name: some-name-vault

Reviewer: it is needed to mention where name is mentioned in CR

      type: Opaque
      stringData:
        keyring_vault.conf: |-
          token = s.VgQvaXl8xGFO1RUxAPbPbsfN
          vault_url = vault-service.vault-service.svc.cluster.local

>           secret_mount_point = secret

Reviewer: ```suggestion
          secret_mount_point = pxc-secret
```

More details on how to install and configure Vault can be found `in the official documentation <https://learn.hashicorp.com/vault?track=getting-started-k8s#getting-started-k8s>`_.

.. _vault-encryption:

Using the encryption
-------------------------------

If using Percona XtraDB Cluster 5.7, you should turn encryption on explicitly
when you create a table or a tablespace. This can be done by adding the
``ENCRYPTION='Y'`` part to your SQL statement, like in the following example:

   .. code:: sql

      CREATE TABLE t1 (c1 INT, PRIMARY KEY pk(c1)) ENCRYPTION='Y';

>       CREATE TABLESPACE foo ADD DATAFILE 'foo.ibd' ENCRYPTION='Y';

Reviewer: should we add a link about encryption in 5.7?

.. _encryption:

Data at rest encryption
************************

`Ful data at rest encryption in Percona XtraDB Cluster <https://www.percona.com/doc/percona-xtradb-cluster/LATEST/management/data_at_rest_encryption.html>`_ is supported by the Operator since version 1.4.0.

..note:: "Data at rest" means inactive data stored as files, database records, etc.

To implement these features, the Operator uses ``keyring_vault`` plugin,
which ships with Percona XtraDB Cluster, and utilizes `HashiCorp Vault <https://www.vaultproject.io/>`_ authentication server.

.. contents:: :local:

.. _install-vault:

Installing Vault
----------------

>  user feel free to setup Vault as he wants, steps 1-4 are only illustrative

Reviewer: typo
```suggestion
```

.. _encryption:

> Data at rest encryption

Reviewer 2: Capitalize in titles.  Should be Data-at-Rest Encryption

Data-at-Rest when written out is typically hyphenated.

************************

> `Ful data at rest encryption in Percona XtraDB Cluster <https://www.percona.com/doc/percona-xtradb-cluster/LATEST/management/data_at_rest_encryption.html>`_ is supported by the Operator since version 1.4.0.

Reviewer 2: "Ful" should be "Full"

> ..note:: "Data at rest" means inactive data stored as files, database records, etc.

Reviewer 2: We could link to an outside resource here such as Wikipedia if we'd like.  Not required though.

To implement these features, the Operator uses ``keyring_vault`` plugin,
which ships with Percona XtraDB Cluster, and utilizes `HashiCorp Vault <https://www.vaultproject.io/>`_ storage for encryption keys.

.. contents:: :local:

.. _install-vault:

> Installing Vault

Reviewer 2: Have these instructions been tested and verified?

      $ kubectl exec -it vault-0 -- /bin/sh
      $ vault login s.VgQvaXl8xGFO1RUxAPbPbsfN
      $ vault secrets enable --version=1 -path=pxc-secret kv

   .. note:: You can also enable audit, which is not mandatory, but useful:

      .. code:: bash

         $ vault audit enable file file_path=/vault/vault-audit.log

2. To enable Vault secret within Kubernetes, create and apply the YAML file,
   as described further.

   2.1. To access the Vault server via HTTP, follow the next YAML file example:

      .. code:: text

         apiVersion: v1
         kind: Secret
         metadata:
           name: some-name-vault
         type: Opaque
         stringData:
           keyring_vault.conf: |-
              token = s.VgQvaXl8xGFO1RUxAPbPbsfN
              vault_url = vault-service.vault-service.svc.cluster.local
              secret_mount_point = pxc-secret

      .. note:: the ``name`` key in the above file should be equal to the
         ``spec.vaultSecretName`` key from the ``deploy/cr.yaml`` configuration
         file.

   2.2. To turn on TLS and access the Vault server via HTTPS, you should do two more things:
      

>       * add one more item to the secret: the contents of the ``ca.cert`` file

Reviewer 2: Can you put an example certificate in the sample YAML?

        with your certificate,
      * store the path to this file in the ``vault_ca`` key.

      .. code:: text

         apiVersion: v1
         kind: Secret
         metadata:
           name: some-name-vault
         type: Opaque
         stringData:
           keyring_vault.conf: |-
             token = = s.VgQvaXl8xGFO1RUxAPbPbsfN
             vault_url = https://vault-service.vault-service.svc.cluster.local
             secret_mount_point = pxc-secret
             vault_ca = /etc/mysql/vault-keyring-secret/ca.cert
           ca.cert: |-

      .. note:: the ``name`` key in the above file should be equal to the
         ``spec.vaultSecretName`` key from the ``deploy/cr.yaml`` configuration
         file.
         
      .. note:: For techincal reasons the ``vault_ca`` key should either exist
         or not exist in the YAML file; commented option like
         ``#vault_ca = ...`` is not acceptable.

More details on how to install and configure Vault can be found `in the official documentation <https://learn.hashicorp.com/vault?track=getting-started-k8s#getting-started-k8s>`_.

> 

Reviewer: please excuse me and revert https://github.com/percona/percona-xtradb-cluster-operator/pull/362/commits/1f83663189ec97dbea77000585da16e63389dea1

.. _encryption:

Data at rest encryption
************************

`Ful data at rest encryption in Percona XtraDB Cluster <https://www.percona.com/doc/percona-xtradb-cluster/LATEST/management/data_at_rest_encryption.html>`_ is supported by the Operator since version 1.4.0.

..note:: "Data at rest" means inactive data stored as files, database records, etc.

To implement these features, the Operator uses ``keyring_vault`` plugin,

> which ships with Percona XtraDB Cluster, and utilizes `HashiCorp Vault <https://www.vaultproject.io/>`_ authentication server.

Reviewer: ```suggestion
which ships with Percona XtraDB Cluster, and utilizes `HashiCorp Vault <https://www.vaultproject.io/>`_ storage for encryption keys.
```

1. First, you should enable secrets within Vault. Get the Vault root token:

   .. code:: bash

      $ cat /tmp/vault-init | jq -r ".root_token"

   The output will be like follows:

   .. code:: text

      s.VgQvaXl8xGFO1RUxAPbPbsfN

   Now login to Vault with this token and enable the "pxc-secret" secrets path:

      $ kubectl exec -it vault-0 -- /bin/sh
      $ vault login s.VgQvaXl8xGFO1RUxAPbPbsfN
      $ vault secrets enable --version=1 -path=pxc-secret kv

   .. note:: You can also enable audit, which is not mandatory, but useful:

      .. code:: bash

         $ vault audit enable file file_path=/vault/vault-audit.log

2. To enable Vault secret within Kubernetes, create and apply the YAML file,
   as described further.

   2.1. To access the Vault server via HTTP, follow the next YAML file example:

      .. code:: text

         apiVersion: v1
         kind: Secret
         metadata:
           name: some-name-vault
         type: Opaque
         stringData:
           keyring_vault.conf: |-
              token = s.VgQvaXl8xGFO1RUxAPbPbsfN
              vault_url = vault-service.vault-service.svc.cluster.local
              secret_mount_point = pxc-secret

>       .. note:: the ``name`` key in the above file should be equal to the

Reviewer: is both 2.1 and 2.2 should have this?

Author: fixed, wrong alignment

Reviewer: still see it only for 2.1
It is actual for both 2.1 and 2.2 sections

         ``spec.vaultSecretName`` key from the ``deploy/cr.yaml`` configuration
         file.

   2.2. To turn on TLS and access the Vault server via HTTPS, you should do two more things:
      
      * add one more item to the secret: the contents of the ``ca.cert`` file
        with your certificate,
      * store the path to this file in the ``vault_ca`` key.

      .. code:: text

         apiVersion: v1
         kind: Secret
         metadata:
           name: some-name-vault
         type: Opaque
         stringData:
           keyring_vault.conf: |-
             token = = s.VgQvaXl8xGFO1RUxAPbPbsfN

>              vault_url = vault-service.vault-service.svc.cluster.local

Reviewer: ```suggestion
             vault_url = https://vault-service.vault-service.svc.cluster.local
```

             secret_mount_point = pxc-secret
             vault_ca = /etc/mysql/vault-keyring-secret/ca.cert
           ca.cert: |-
             ..........

      .. note:: For techincal reasons the ``vault_ca`` key should either exist
         or not exist in the YAML file; commented option like
         ``#vault_ca = ...`` is not acceptable.

More details on how to install and configure Vault can be found `in the official documentation <https://learn.hashicorp.com/vault?track=getting-started-k8s#getting-started-k8s>`_.

.. _vault-encryption:

Using the encryption
--------------------

If using Percona XtraDB Cluster 5.7, you should turn encryption on explicitly
when you create a table or a tablespace. This can be done by adding the
``ENCRYPTION='Y'`` part to your SQL statement, like in the following example:

   .. code:: sql

      CREATE TABLE t1 (c1 INT, PRIMARY KEY pk(c1)) ENCRYPTION='Y';
      CREATE TABLESPACE foo ADD DATAFILE 'foo.ibd' ENCRYPTION='Y';

.. note:: See more details on encryption in Percona XtraDB Cluster 5.7 `here <https://www.percona.com/doc/percona-xtradb-cluster/5.7/management/data_at_rest_encryption.html>`_.

> If using Percona XtraDB Cluster 8.0, the encryption is turned on by default.

Reviewer: we can remove whole "Using the encryption" section, as soon as @tduzan provide list of options for my.cnf
